### PR TITLE
Migrate Package revision diff to bootstrap

### PR DESCRIPTION
**Pull request by ChrisBr**
To continue our efforts to migrate the whole frontend to bootstrap.
https://github.com/openSUSE/open-build-service/wiki/How-to-work-in-Bootstrap-%28new-UI%29

## Not expanded diffs
selection_026

## Small expanded diff
selection_027

## Large expanded diff
selection_028

### CODE REVIEW

**Comment by ChrisBr** [comment]
I don't like the ``^`` for the jump button too much (jumping from diff to diff):

selection_029

It was like this before and I think the functionality is crucial for some people. Does anyone has a better idea as just ``^``?

File: src/api/app/controllers/webui/package_controller.rb [hunk 1]
```diff
@@ -484,6 +484,19 @@ def rdiff
     @files = filenames['files']
     @not_full_diff = @files.any? { |file| file[1]['diff'].try(:[], 'shown') }
     @filenames = filenames['filenames']
+
+    # TODO: moved from the old view, needs refactoring
```
Comment (Ana06) [inline src/api/app/controllers/webui/package_controller.rb]: it was only copied... should it be removed from the old view if it is now in the controller? :thinking:

Comment (ChrisBr) [reply src/api/app/controllers/webui/package_controller.rb]: The old view gets anyway removed in a couple of weeks, no need to clean up IMHO...

**Comment by DavidKang** [comment]
@ChrisBr, if we put `jump` instead of `view file` and we add an arrow in the correct direction because is jumping from top to bottom, right?

**Comment by ChrisBr** [comment]
> if we put jump instead of view file and we add an arrow in the correct direction because is jumping from top to bottom, right?

Nope, it is jumping up. And there are two links in the button group:
* View file links to a new page showing the whole file
* ^ jumps UP to the next diff (from bottom to top)

**Comment by DavidKang** [comment]
better than `^` is using [chevron-up](https://fontawesome.com/icons/chevron-up?style=solid)

**Comment by ChrisBr** [comment]
I added some files here:

https://obs-reviewlab.opensuse.org/chrisbr-bootstraprdiff/package/rdiff/home:Admin/ruby?linkrev=base&rev=3

**Comment by ChrisBr** [comment]
> better than ^ is using chevron-up

:heart:

**Comment by DavidKang** [comment]
And we can also add title attribute if is needed

**Comment by ChrisBr** [comment]
selection_030

Looks already better :+1:

File: src/api/app/controllers/webui/package_controller.rb (hunk 1 above)
Comment (Ana06) [reply src/api/app/controllers/webui/package_controller.rb]: ok

File: src/api/app/helpers/webui/package_helper.rb [hunk 2]
```diff
@@ -88,4 +88,28 @@ def repo_type_and_priority(repository)
...
+
+  def expand_diff?(filename, state)
+    state != 'Deleted' &&
+      !filename.include?('/') &&
+      (filename == '_patchinfo' || filename.ends_with?('.spec') || filename.ends_with?('.changes'))
+  end
+
+  def calculate_revision_on_state(revision, state)
+    result = revision.to_i
+    state == 'deleted' && result > 0 ? result - 1 : result
```
Comment (Ana06) [inline src/api/app/helpers/webui/package_helper.rb]: [nitpick] I find it more readable like:

```ruby
result = revision.to_i
result -= 1 if state == 'deleted'
max(0,result)
```

**Comment by ChrisBr** [comment]
> Also, if you don't click exactly in the ^ but a little bit to the side, it doesn't work and it is really confusing.

Changing the ``^`` to font awesome broke :cry: Fixed...

File: src/api/app/views/webui2/webui/package/_revision_diff_detail.html.haml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    = filename
+    %span.badge{ class: badge_for_diff_state(file['state']) }= file['state'].capitalize
+    .btn-group.float-right{ role: :group }
+      - unless @linkinfo
+        = link_to_unless(Package.is_binary_file?(filename) || filename.include?('/'),
+        'View file',
+        package_view_file_path(project: project, package: package, filename: filename, rev: calculate_revision_on_state(revision, file['state'])),
+        class: 'btn btn-sm btn-outline-secondary', title: 'View the whole file')
+      - if index.positive?
+        %a.btn.btn-sm.btn-outline-secondary.fas.fa-chevron-up{ href: "#revision_details_#{index - 1}" }
```
Comment (Ana06) [inline src/api/app/views/webui2/webui/package/_revision_diff_detail.html.haml]: This is what makes it so weird! I was getting really confused :eyes: This should be `index` or [up](https://github.com/openSUSE/open-build-service/pull/5943/files#diff-6889ea0ae17f916c39b255e1a700bbd0R1) `index-1`

File: src/api/app/assets/javascripts/webui2/package.js [hunk 4]
```diff
@@ -0,0 +1,8 @@
+$(document).ready(function(){
+  $('details.details-with-codemirror').on('click', function () {
+    var editor = $(this).find('.CodeMirror')[0].CodeMirror;
+    setTimeout(function() {
+      editor.refresh();
+    },1);
```
Comment (bgeuken) [inline src/api/app/assets/javascripts/webui2/package.js]: newline is missing here^^

**Comment by ChrisBr** [comment]
@DavidKang I had
```
%a.btn.btn-sm.btn-outline-secondary.fas.fa-chevron-up{ style: 'height: 16px, width: 16px' }
```

and it did not work. Was it that what you meant?

**Comment by dmarcoux** [comment]
@ChrisBr The chevron icon (`%i.fas.fa-chevron-down`) needs to be inside the `%a`.

File: src/api/app/views/webui2/webui/package/_revision_diff_detail.html.haml [hunk 5]
```diff
@@ -0,0 +1,18 @@
+%details.card.details-with-codemirror{ open: expand_diff?(filename, file['state']), id: "revision_details_#{index}" }
+  %summary.card-header
+    = filename
+    %span.badge{ class: badge_for_diff_state(file['state']) }= file['state'].capitalize
+    .btn-group.float-right{ role: :group }
+      - if index.positive?
+        %a.btn.btn-sm.btn-outline-secondary{ href: "#revision_details_#{index - 1}", title: 'Go up to the next diff' }
```
Comment (dmarcoux) [inline src/api/app/views/webui2/webui/package/_revision_diff_detail.html.haml]: The title should be `Go up to the previous diff`

**Comment by DavidKang** [comment]
@ChrisBr, I created to commits, @Ana06 told me about the bug in Chrome/Chromium. This should fix the issue.

File: src/api/spec/helpers/webui/package_helper_spec.rb [hunk 6]
```diff
@@ -253,4 +253,25 @@
     it { expect(uploadable?('image.vhdfixed.xz', 'i386')).to be_falsy }
     it { expect(uploadable?('apache2.rpm', 'x86_64')).to be_falsy }
   end
+
+  describe '#expand_diff?' do
+    it { expect(expand_diff?('ctris.spec', 'added')).to be true }
+    it { expect(expand_diff?('_patchinfo', 'added')).to be true }
+    it { expect(expand_diff?('ctris.changes', 'added')).to be true }
+    it { expect(expand_diff?('ctris.changes', 'deleted')).to be false }
+    it { expect(expand_diff?('/foo/bar/test.txt', 'added')).to be false }
```
Comment (saraycp) [inline src/api/spec/helpers/webui/package_helper_spec.rb]: :thinking: I would use `be_true` and `be_false` in the last 5 lines

File: src/api/app/helpers/webui/webui_helper.rb [hunk 7]
```diff
@@ -194,6 +194,7 @@ def elide_two(text1, text2, overall_length = 40, mode = :middle)
   end
 
   def force_utf8_and_transform_nonprintables(text)
+    return '' unless text.present?
```
[the remark this answers is not in the document]
Comment (ChrisBr) [reply src/api/app/helpers/webui/webui_helper.rb]: Already fixed by @mdeniz, dropping this commit!

File: src/api/spec/helpers/webui/package_helper_spec.rb (hunk 6 above)
Comment (ChrisBr) [reply src/api/spec/helpers/webui/package_helper_spec.rb]: ``
expected true to respond to `true?` or perhaps you meant `be true` or `be_truthy`
``

https://eddyluten.com/rspec-be_truthy-exists-or-be-true

Comment (saraycp) [reply src/api/spec/helpers/webui/package_helper_spec.rb]: I meant this:
https://relishapp.com/rspec/rspec-expectations/v/2-2/docs/matchers/be-matchers
What do you think?

Comment (saraycp) [reply src/api/spec/helpers/webui/package_helper_spec.rb]: Oh, I see! You need to check exactly `true` and `false`. So, forget my comment!

**Pull request merged**